### DISCUSSION TITLE
How to record rtp packets from server, need help!

### DISCUSSION THREAD

**Discussion by aijyo** [paullouisageneau/libdatachannel#1070] [Q&A]
I've seen another question about recording（https://github.com/paullouisageneau/libdatachannel/discussions/712） and read datachannel-media-receiver example， But still need help。

  I cannot understand why the Client needs to create a (rtc::Description::Direction::RecvOnly) track by itself:
![image](https://github.com/paullouisageneau/libdatachannel/assets/7956449/b2229a0e-3343-4816-8f6e-c7df24610d15)
  Because I don't know how many audio and video streams the other side has. My understanding is to get how many tracks there are in the PeerConnectionton onTrack event (my code is below. ), but in addition to receiving the ontrack event, the onMessage event of the track was not triggered.
  Now I want to know whether this library can realize the requirement of recording RTP packets. If so, can you tell me how to obtain the RTP packets， thank you so much~

my PeerConnection track code is :
![image](https://github.com/paullouisageneau/libdatachannel/assets/7956449/dc216d34-0d5a-4955-a574-ec957d834898)

**Comment by paullouisageneau** [comment]
> I cannot understand why the Client needs to create a (rtc::Description::Direction::RecvOnly) track by itself:
image
Because I don't know how many audio and video streams the other side has.

There is no concept of what tracks a side has, this is just a negotiation process and it's entirely up to the application. One side has to offer the tracks setup it wants, and the answering side has to accept or reject it, irrelevant of the actual data flow direction. Here the example wants to receive a single track, and it wants it to be h264 and not another codec, so that's what it offers.

> My understanding is to get how many tracks there are in the PeerConnectionton onTrack event (my code is below. ), but in addition to receiving the ontrack event, the onMessage event of the track was not triggered.
> Now I want to know whether this library can realize the requirement of recording RTP packets. If so, can you tell me how to obtain the RTP packets， thank you so much~

The `onMessage` won't be triggered in your code because you drop the shared_ptr to the track at the end of the `onTrack` callback, so the object is destroyed. You need to store the shared_ptr somewhere. It can be either saved externally, or captured in its own callback for instance, in which the object will be destroyed when you close the track:
```cpp
t->onOpen([mid, t]() {
    [...]
});
```
